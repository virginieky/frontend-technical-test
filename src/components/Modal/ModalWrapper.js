import styled from 'styled-components';
import { Modal } from 'reactstrap';

const ModalWrapper = styled(Modal)`
  button.close {
    background: white;
    border: 0;
    font-size: 20px;
    outline: 0;
  }
`;

export default ModalWrapper;
