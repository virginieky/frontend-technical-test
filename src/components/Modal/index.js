import React, { useState } from 'react';
import PropTypes from 'prop-types';

import {
  Button,
  Form,
  FormGroup,
  Input,
  Label,
  Modal as ReactstrapModal,
  ModalHeader,
  ModalBody,
  ModalFooter,
} from 'reactstrap';

import {
  getCreateButtonClick,
  getModalToggle,
  getRecipientChange,
} from './utils';

const Modal = ({ onCreate, options }) => {
  const [modal, setModal] = useState(false);
  const [recipient, setRecipient] = useState('');
  const isSubmitButtonDisabled = !recipient;
  const toggle = getModalToggle({ setModal });

  return (
    <div>
      <Button color='danger' onClick={toggle}>
        New conversation
      </Button>
      <ReactstrapModal isOpen={modal} toggle={toggle}>
        <ModalHeader toggle={toggle}>Create new conversation</ModalHeader>
        <ModalBody>
          {options.length > 0 ? (
            <Form>
              <FormGroup>
                <Label for='recipient'>Choose a contact</Label>
                <Input
                  type='select'
                  name='recipient'
                  id='recipient'
                  value={recipient}
                  onChange={getRecipientChange({ setRecipient })}
                >
                  <option value=''>Please choose a contact</option>
                  {options.map((option) => {
                    return (
                      <option key={option.id} value={option.id}>
                        {option.nickname}
                      </option>
                    );
                  })}
                </Input>
              </FormGroup>
            </Form>
          ) : (
            <p>You don't have any contact for the moment!</p>
          )}
        </ModalBody>
        <ModalFooter>
          {options.length > 0 && (
            <Button
              disabled={isSubmitButtonDisabled}
              color='primary'
              onClick={getCreateButtonClick({
                options,
                recipient,
                onCreate,
                toggle,
              })}
            >
              Create
            </Button>
          )}
          <Button color='secondary' onClick={toggle}>
            Cancel
          </Button>
        </ModalFooter>
      </ReactstrapModal>
    </div>
  );
};

Modal.defaultProps = {
  onCreate: () => {},
  options: [],
};

Modal.propTypes = {
  onCreate: PropTypes.func,
  options: PropTypes.array,
};

export default Modal;
