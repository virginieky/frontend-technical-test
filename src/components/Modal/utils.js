export const getCreateButtonClick =
  ({ options, recipient, onCreate, toggle }) =>
  () => {
    const recipientId = parseInt(recipient);
    const selectedOption = options.find((option) => option.id === recipientId);

    onCreate({
      recipientId,
      recipientNickname: selectedOption.nickname,
    });
    toggle();
  };

export const getModalToggle =
  ({ setModal }) =>
  () => {
    setModal((prevState) => !prevState);
  };

export const getRecipientChange =
  ({ setRecipient }) =>
  ({ target: { value } }) => {
    setRecipient(value);
  };
