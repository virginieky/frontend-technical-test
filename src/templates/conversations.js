const conversations = [
  {
    id: 1,
    recipientId: 2,
    recipientNickname: 'Jeremie',
    senderId: 1,
    senderNickname: 'Thibaut',
    lastMessageTimestamp: 1625637849,
  },
  {
    id: 2,
    recipientId: 3,
    recipientNickname: 'Patrick',
    senderId: 1,
    senderNickname: 'Thibaut',
    lastMessageTimestamp: 1620284667,
  },
  {
    id: 3,
    recipientId: 1,
    recipientNickname: 'Thibaut',
    senderId: 4,
    senderNickname: 'Elodie',
    lastMessageTimestamp: 1625648667,
  },
  {
    recipientId: 5,
    recipientNickname: 'Maeko',
    senderId: 1,
    senderNickname: 'Thibaut',
    lastMessageTimestamp: 1632230669743,
    id: 4,
  },
  {
    recipientId: 5,
    recipientNickname: 'Maeko',
    senderId: 1,
    senderNickname: 'Thibaut',
    lastMessageTimestamp: 1632230756751,
    id: 5,
  },
  {
    recipientId: 6,
    recipientNickname: 'Duarte',
    senderId: 1,
    senderNickname: 'Thibaut',
    lastMessageTimestamp: 1632230824677,
    id: 6,
  },
];

export default conversations;
