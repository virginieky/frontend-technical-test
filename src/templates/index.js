export { default as conversations } from './conversations';
export { default as messages } from './messages';
export { default as users } from './users';
