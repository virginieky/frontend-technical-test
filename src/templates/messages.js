const messages = [
  {
    id: 1,
    conversationId: 1,
    timestamp: 1625637849,
    authorId: 1,
    body: "Bonjour c'est le premier message de la première conversation",
  },
  {
    id: 2,
    conversationId: 1,
    timestamp: 1625637867,
    authorId: 1,
    body: "Bonjour c'est le second message de la première conversation",
  },
  {
    id: 3,
    conversationId: 1,
    timestamp: 1625648667,
    authorId: 2,
    body: "Bonjour c'est le troisième message de la première conversation",
  },
  {
    id: 4,
    conversationId: 2,
    timestamp: 1620284667,
    authorId: 2,
    body: "Bonjour c'est le premier message de la seconde conversation",
  },
];

export default messages;
