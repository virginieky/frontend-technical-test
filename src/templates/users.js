const users = [
  {
    id: 1,
    nickname: 'Thibaut',
    token: 'xxxx',
  },
  {
    id: 2,
    nickname: 'Jeremie',
    token: 'xxxx',
  },
  {
    id: 3,
    nickname: 'Patrick',
    token: 'xxxx',
  },
  {
    id: 4,
    nickname: 'Elodie',
    token: 'xxxx',
  },
  {
    id: 5,
    nickname: 'Maeko',
    token: 'xxxx',
  },
  {
    id: 6,
    nickname: 'Duarte',
    token: 'xxxx',
  },
];

export default users;
