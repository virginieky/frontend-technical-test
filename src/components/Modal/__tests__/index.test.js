import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { users } from '../../../templates';
import Modal from '../index';

describe('components | Modal', () => {
  const defaultProps = {
    options: users,
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<Modal {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display the modal on new button click', async () => {
    render(<Modal {...defaultProps} />);
    const newConversationButton = screen.getByRole('button', {
      name: /New conversation/i,
    });
    await userEvent.click(newConversationButton);

    expect(screen.getByText('Create new conversation')).toBeInTheDocument();
  });

  it('should use onCreate defaultProps', () => {
    const {
      defaultProps: { onCreate },
    } = Modal;

    expect(onCreate).toBeDefined();
    expect(onCreate({ preventDefault: jest.fn() })).toBe(undefined);
  });
});
