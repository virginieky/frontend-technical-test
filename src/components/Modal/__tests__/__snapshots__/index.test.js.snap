// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components | Modal should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <button
      class="btn btn-danger"
      type="button"
    >
      New conversation
    </button>
  </div>
</DocumentFragment>
`;
